### PR TITLE
copy-tracking: improve --summary and add --stat

### PR DESCRIPTION
**Pull request by fowles**
# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [x] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

**Comment by ilyagr** [comment]
How is this related to #4118 (and is it)? I haven't been following that discussion as closely as I would have liked, so I'm wondering whether to look there or here first when I finally get to trying to understand what you are doing (no need to wait for me though).

**Comment by fowles** [comment]
> How is this related to #4118 (and is it)? I haven't been following that discussion as closely as I would have liked, so I'm wondering whether to look there or here first when I finally get to trying to understand how copy-tracking works (no need to wait for me though).

it builds on it.  You only need to look at the last commit on this one.

File: CHANGELOG.md [hunk 1]
```diff
@@ -14,8 +14,11 @@ to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ### New features
 
+<<<<<<< Conflict 1 of 1
++++++++ Contents of side #1
```
Comment (PhilipMetzger) [inline CHANGELOG.md:18]: Nit: Remove the partially commited conflicted resolution.

File: cli/src/text_util.rs [hunk 2]
```diff
@@ -35,6 +37,33 @@ pub fn split_email(email: &str) -> (&str, Option<&str>) {
...
+        .last()
+        .map(|((s, _), (t, _))| (s, t))
+        .unwrap_or((0, 0));
+    let suffix = UnicodeSegmentation::split_word_bound_indices(source)
+        .rev()
+        .zip(UnicodeSegmentation::split_word_bound_indices(target).rev())
+        .take_while(|((_, s), (_, t))| s == t)
+        .last()
+        .map(|((s, _), (t, _))| (s, t))
+        .unwrap_or((source.len(), target.len()));
```
Comment (PhilipMetzger) [inline cli/src/text_util.rs:56]: Nit: the whole block needs readable names, iiuc `s` => `source`, `t` => `target`

File: cli/src/text_util.rs [hunk 3]
```diff
@@ -35,6 +37,33 @@ pub fn split_email(email: &str) -> (&str, Option<&str>) {
     }
 }
 
+pub fn render_copied_path(source: &str, target: &str) -> String {
+    if source == target {
+        return target.into();
+    }
```
Comment (PhilipMetzger) [inline cli/src/text_util.rs:43]: Nit: move this into the caller for now.

**Review by PhilipMetzger — CHANGES_REQUESTED** [review]
This is great, there are just some minor nits.

File: cli/src/text_util.rs (hunk 2 above)
Comment (fowles) [reply cli/src/text_util.rs:56]: done

File: cli/src/text_util.rs (hunk 3 above)
Comment (fowles) [reply cli/src/text_util.rs:43]: I don't like that.  As it stands you can always pass strings to this and it will generate the right thing.  If you don't have this check, it will do something weird for the of equal strings.

File: CHANGELOG.md (hunk 1 above)
Comment (fowles) [reply CHANGELOG.md]: done

File: cli/src/text_util.rs [hunk 4]
```diff
@@ -629,4 +658,30 @@ mod tests {
...
+            render_copied_path("one/two/three", "one/three"),
+            "one/{two => }/three"
+        );
+        assert_eq!(
+            render_copied_path("one/two", "one/four"),
+            "one/{two => four}"
+        );
+        assert_eq!(render_copied_path("two", "four"), "{two => four}");
+        assert_eq!(render_copied_path("file1", "file2"), "{file1 => file2}");
+    }
```
Comment (PhilipMetzger) [inline cli/src/text_util.rs:686]: Nit: Missing test for the equal path handling, if you want to keep it.

File: cli/src/diff_util.rs [hunk 5]
```diff
@@ -1244,21 +1254,28 @@ pub fn show_diff_stat(
     display_width: usize,
 ) -> Result<(), DiffRenderError> {
     let mut stats: Vec<DiffStat> = vec![];
+    let mut unresolved_renames = HashSet::<String>::new();
```
Comment (PhilipMetzger) [inline cli/src/diff_util.rs:1257]: This is a stylistic nit, this can just be `HashSet::new` as Rust will infer it on line 1276.

File: cli/src/text_util.rs (hunk 3 above)
Comment (PhilipMetzger) [reply cli/src/text_util.rs:43]: Basically the caller can handle the no-op or it should be documented somewhere, as it currently is a bit magic.

**Review by PhilipMetzger — APPROVED** [review]
LG, maybe Yuya or Martin should take a look too

File: cli/src/text_util.rs [hunk 6]
```diff
@@ -35,6 +37,33 @@ pub fn split_email(email: &str) -> (&str, Option<&str>) {
...
+        .last()
+        .map(|((source_index, _), (target_index, _))| (source_index, target_index))
+        .unwrap_or((0, 0));
+    let suffix = UnicodeSegmentation::split_word_bound_indices(source)
+        .rev()
+        .zip(UnicodeSegmentation::split_word_bound_indices(target).rev())
+        .take_while(|((_, source_word), (_, target_word))| source_word == target_word)
+        .last()
+        .map(|((source_index, _), (target_index, _))| (source_index, target_index))
+        .unwrap_or((source.len(), target.len()));
```
Comment (yuja) [inline cli/src/text_util.rs:56]: Is it intentional that `file-1 => file-2` is rendered as `file-{1 => 2}`? I personally find it's hard to follow, and would rather want to split by path separator, not at word boundary.

BTW, a generic implementation of this function would be something like
```rust
Diff::for_tokenizer([source, target], find_path_component_ranges)
    .map(|hunk| match hunk {
        DiffHunk::Matching(common) => common,
        DiffHunk::Different([source, target]) => "{source} => {target}", // XXX handle trailing sep?
     })
     .collect()
```

File: cli/src/commit_templater.rs [hunk 7]
```diff
@@ -1394,14 +1394,17 @@ fn builtin_tree_diff_methods<'repo>() -> CommitTemplateBuildMethodFnMap<'repo, T
             let path_converter = language.path_converter;
             let template = self_property
                 .map(move |diff| {
+                    let from_tree = diff.from_tree.clone();
                     let to_tree = diff.to_tree.clone();
-                    diff.into_formatted(move |formatter, _store, tree_diff| {
+                    let matcher = diff.matcher.clone();
+                    diff.into_formatted(move |formatter, _store, _tree_diff| {
```
Comment (yuja) [inline cli/src/commit_templater.rs:1400]: nit: can you leave a TODO comment?
`// TODO: don't pass separate copies of from_tree/to_tree/matcher`

File: cli/src/text_util.rs (hunk 3 above)
Comment (fowles) [reply cli/src/text_util.rs:43]: added docs, there was already a test for this behavior.

File: cli/src/diff_util.rs (hunk 5 above)
Comment (fowles) [reply cli/src/diff_util.rs:1257]: done

File: cli/src/commit_templater.rs (hunk 7 above)
Comment (fowles) [reply cli/src/commit_templater.rs:1400]: Done

File: cli/src/text_util.rs (hunk 6 above)
Comment (fowles) [reply cli/src/text_util.rs:56]: That is a good point.  I have moved this into `repo_path` so it sits with the other format function for changing paths into UI strings.  I have also update the implementation to be component based (which means we no long have a dep on unicode).

File: lib/src/repo_path.rs [hunk 8]
```diff
@@ -502,6 +503,70 @@ impl RepoPathUiConverter {
         }
     }
 
+    /// Format a copy from `source` to `target` for display in the UI by
+    /// extracting common components and producing something like
+    /// "common/prefix/{source => target}/common/suffix".
+    ///
+    /// If `source == target`, returns `format_file_path(source)`.
```
Comment (martinvonz) [inline lib/src/repo_path.rs:510]: I'm not sure this belongs in the lib crate. It feels a bit specific to a CLI. It seems likely that e.g. a GUI or Web UI would want to display it differently. Actually, even the CLI may want to style the renamed part differently from the rest of the string. So I feel like this should be made more generic or be moved into CLI. It could be made more generic by returning the `(prefix, (before, after), suffix)` or something like that. Maybe a proper struct or enum is better (an enum could have a `Unchanged(String)` variant, but it could instead be a struct wrapped in an `Option`). What do you think?

Comment (fowles) [reply lib/src/repo_path.rs:510]: do you have a suggestion for where to move it?  One of the reasons it fits here in my mind is that `format_file_path` is also doing a bunch of UI specific work and they do share behavior that core part of things.

Comment (martinvonz) [reply lib/src/repo_path.rs:510]: I think `format_file_path()` is generic enough that it makes sense in a web UI too, for example, but a web UI would probably not want to display the `common/prefix/{source => target}/common/suffix` form. Maybe `diff_util.rs` is a good place since this way of displaying paths is specific to diffs.

Comment (martinvonz) [reply lib/src/repo_path.rs:510]: Oh, but maybe the use of `RepoPathUiConverter` means that it should be here. Let me take a look at the code to see how this works.

File: lib/src/repo_path.rs [hunk 9]
```diff
@@ -858,4 +923,29 @@ mod tests {
...
+                .replace("\\", "/")
+        };
+
+        assert_eq!(format("one/two/three", "one/two/three"), "one/two/three");
+        assert_eq!(format("two/three", "four/three"), "{two => four}/three");
+        assert_eq!(
+            format("one/two/three", "one/four/three"),
+            "one/{two => four}/three"
+        );
+        assert_eq!(format("one/two/three", "one/three"), "one/{two => }/three");
```
Comment (martinvonz) [inline lib/src/repo_path.rs:945]: Could you add tests like the following too?
```
        assert_eq!(format("one/two", "one/two/three"), "one/{two => two/three}");
        assert_eq!(format("one/two", "zero/one/two"), "{one/ => zero/one/}two");
        assert_eq!(format("one/two/three", "one/two"), "one/{two/three => three}");
        assert_eq!(format("zero/one/two", "one/two"), "{zero/one/ => one/}two");
        assert_eq!(format("one/two", "one/two/three/one/two"), "one/{two => two/three/one/two}");
```

I think that's how I would expect the result to look but I'm not sure. You can check what git does if you feel like it.

File: lib/src/repo_path.rs (hunk 8 above)
Comment (martinvonz) [reply lib/src/repo_path.rs:510]: I'm still not sure where this code should live but we can always move it later. Let's leave it here for now unless someone else has a good idea for how to structure it. Sorry about my indecision.

File: lib/src/repo_path.rs (hunk 9 above)
Comment (fowles) [reply lib/src/repo_path.rs:945]: done
